File: workflow/java/src/main/java/io/dapr/quickstarts/workflows/OrderProcessingWorkflow.java
package io.dapr.quickstarts.workflows;

import org.springframework.stereotype.Service;
import io.dapr.workflows.Workflow;
import io.dapr.workflows.WorkflowStub;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.dapr.quickstarts.workflows.models.*;
import io.dapr.quickstarts.workflows.activities.*;

@Service
public class OrderProcessingWorkflow implements Workflow {

  private static final Logger logger = LoggerFactory.getLogger(OrderProcessingWorkflow.class);

  @Override
  public WorkflowStub create() {
    return ctx -> {
      String orderId = ctx.getInstanceId();
      OrderPayload order = ctx.getInput(OrderPayload.class);
      OrderResult orderResult = new OrderResult(false, "");

      // Notify the user that an order has come through
      Notification notification = new Notification();
      notification.setMessage("Received order " + orderId + " for " + order.getQuantity() + " " + order.getName());
      ctx.callActivity(NotifyActivity.class.getCanonicalName(), notification).await();

      // Determine if there is enough of the item available for purchase by checking
      // the inventory
      InventoryRequest inventoryRequest = new InventoryRequest();
      inventoryRequest.setRequestId(orderId);
      inventoryRequest.setItemName(order.getName());
      inventoryRequest.setQuantity(order.getQuantity());
      InventoryResult inventoryResult = ctx.callActivity(ReserveInventoryActivity.class.getCanonicalName(), inventoryRequest, InventoryResult.class).await();

      // If there is insufficient inventory, fail and let the user know
      if (!inventoryResult.isSuccess()) {
        notification.setMessage("Insufficient inventory for " + order.getName());
        orderResult.setMessage("Order failed due to insufficient inventory");
        ctx.callActivity(NotifyActivity.class.getCanonicalName(), notification).await();
        ctx.complete(orderResult);
        return;
      }

      // There is enough inventory available so the user can purchase the item(s).
      // Process their payment
      PaymentRequest paymentRequest = new PaymentRequest();
      paymentRequest.setRequestId(orderId);
      paymentRequest.setItemName(order.getName());
      paymentRequest.setQuantity(order.getQuantity());
      boolean isOK = ctx.callActivity(ProcessPaymentActivity.class.getCanonicalName(), paymentRequest, boolean.class).await();
      if (!isOK) {
        notification.setMessage("Order " + orderId + " Failed! You are now getting a refund");
        orderResult.setMessage("Order failed during payment processing");
        ctx.callActivity(NotifyActivity.class.getCanonicalName(), notification).await();
        ctx.complete(orderResult);
        return;
      }

      // Update the inventory
      inventoryResult = ctx.callActivity(UpdateInventoryActivity.class.getCanonicalName(), inventoryRequest, InventoryResult.class).await();
      if (!inventoryResult.isSuccess()) {
        // If there is an error updating the inventory, let the user know
        notification.setMessage("Order " + orderId + " Failed! You are now getting a refund");
        orderResult.setMessage("Order failed during inventory update");
        ctx.callActivity(NotifyActivity.class.getCanonicalName(), notification).await();
        ctx.complete(orderResult);
        return;
      }

      // Let user know their order was processed
      notification.setMessage("Order " + orderId + " has completed!");
      ctx.callActivity(NotifyActivity.class.getCanonicalName(), notification).await();

      // Complete the workflow with order result is processed
      orderResult.setProcessed(true);
      orderResult.setMessage("Order has completed!");
      ctx.complete(orderResult);
    };
  }
}

File: workflow/csharp/OrderProcessingWorkflow.cs
using System.Threading.Tasks;
using Dapr.Workflow;
using WorkflowApp.Activities;
using WorkflowApp.Models;

namespace WorkflowApp.Workflows
{
    public class OrderProcessingWorkflow : Workflow<OrderPayload, OrderResult>
    {
        public override async Task<OrderResult> RunAsync(WorkflowContext context, OrderPayload order)
        {
            string orderId = context.InstanceId;

            // Notify the user that an order has come through
            await context.CallActivityAsync(
                nameof(NotifyActivity),
                new Notification($"Received order {orderId} for {order.Quantity} {order.Name}"));

            // Determine if there is enough of the item available for purchase by checking the inventory
            InventoryResult result = await context.CallActivityAsync<InventoryResult>(
                nameof(ReserveInventoryActivity),
                new InventoryRequest(orderId, order.Name, order.Quantity));

            // If there is insufficient inventory, fail and let the user know 
            if (!result.Success)
            {
                await context.CallActivityAsync(
                    nameof(NotifyActivity),
                    new Notification($"Insufficient inventory for {order.Name}"));

                context.SetCustomStatus("Stopped order process due to insufficient inventory");

                return new OrderResult(false, "Order failed due to insufficient inventory");
            }

            // There is enough inventory available so the user can purchase the item(s). Process their payment
            await context.CallActivityAsync(
                nameof(ProcessPaymentActivity),
                new PaymentRequest(orderId, order.Name, order.Quantity));

            try
            {
                // Update the inventory
                await context.CallActivityAsync(
                    nameof(UpdateInventoryActivity),
                    new InventoryRequest(orderId, order.Name, order.Quantity));
            }
            catch
            {
                await context.CallActivityAsync(
                    nameof(NotifyActivity),
                    new Notification($"Order {orderId} Failed! You are now getting a refund"));

                context.SetCustomStatus("Stopped order process due to error in inventory update");

                return new OrderResult(false, "Order failed during inventory update");
            }

            // Let them know their payment was processed
            await context.CallActivityAsync(
                nameof(NotifyActivity),
                new Notification($"Order {orderId} has completed!"));

            // End the workflow with a success result
            return new OrderResult(true, "Order has completed!");
        }
    }
}

